framework/sandstone_thread.h

```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright 2023 Intel Corporation.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef SANDSTONE_THREAD
+#define SANDSTONE_THREAD
+
+#include <functional>
+
+#include <pthread.h>
+#include <stdint.h>
+
+struct SandstoneTestThread
+{
+    using RunnerFunction = uintptr_t (int);
+    void start(RunnerFunction *, int cpu);
+    uintptr_t join();
+
+    pthread_t thread;
+    RunnerFunction *target;
+    int start_cpu;
```

Comment:
The class is setting `thread_num` so it needs to store somewhere. Since it is storing, I decided to just as well pass it to the `runner` function, as that costs nothing. And we need the target function pointer, so we need per-thread-object storage anyway, so again cost nothing.

This is also designed this way so that I could, if needed in the future, make a next-level of thread starting. That is, instead of starting up to 64 threads sequentially (32-core hyperthreaded slice), I could start 8 and each of those 8 starts another 7 before calling the target function. This would make an 8-socket 60-core Sapphire Rapids start 16 children, each starting 8 threads, and each thread another 7 (except the last that would start only 3 more), so the longest path is 16+7+8 `clone()` system calls (31, very close to the ideal of √960 = 30.98) instead of the 16+60 of the current branch and 960 in the main branch.

My current benchmarks are inconclusive whether this is necessary. The stack pre-allocation (this PR) made the thread-starting very fast. And yet those benchmarks are showing 12-15 ms overhead; I don't know for sure where the time is going.

Reply:
Ah, it might come in handy for `isolate_socket` and `fullsystem` mode tests, because then we aren't going to be limited to 64 threads. On a 60-core SPR that's 120 threads for each socket, so doing 15*8 would be an improvement, and especially so for the tests that may still need to be full-system.